cyto_dl/nn/vits/jepa.py

```diff
@@ -0,0 +1,325 @@
+        upsample_factor:Optional[List[int]] = [2.6134, 2.5005, 2.5005]
+            Upsampling factor for each dimension (ZYX) order. Default is AICS 20x to 100x object upsampling
+        encoder_ckpt: Optional[str]=None
+            Path to pretrained ViT backbone checkpoint
+        """
+        super().__init__()
+        assert spatial_dims in (2, 3)
+        if isinstance(num_patches, int):
+            num_patches = [num_patches] * spatial_dims
+        if isinstance(patch_size, int):
+            patch_size = [patch_size] * spatial_dims
+        assert len(num_patches) == spatial_dims
+        assert len(patch_size) == spatial_dims
+
+        self.encoder = JEPAEncoder(
+            spatial_dims=spatial_dims,
+            num_patches=num_patches,
+            patch_size=patch_size,
+            emb_dim=emb_dim,
+            **encoder_kwargs,
+        )
+        if encoder_ckpt is not None:
+            model = torch.load(encoder_ckpt, map_location="cuda:0")
+            enc_state_dict = {
+                k.replace("backbone.encoder.", ""): v
+                for k, v in model["state_dict"].items()
+                if "encoder" in k and "intermediate" not in k
+            }
+            self.encoder.load_state_dict(enc_state_dict, strict=False)
+
+        if freeze_encoder:
+            for name, param in self.encoder.named_parameters():
+                # allow different weighting of internal activations for finetuning
+                param.requires_grad = "intermediate_weighter" in name
+
+        self.decoder = torch.nn.Sequential(
+            torch.nn.Linear(emb_dim, 16 * emb_dim),
+            Rearrange(
+                "b (n_patch_z n_patch_y n_patch_x) c -> b c n_patch_z n_patch_y n_patch_x",
+                n_patch_z=num_patches[0],
+                n_patch_y=num_patches[1],
+                n_patch_x=num_patches[2],
+            ),
+            UnetResBlock(
+                spatial_dims=spatial_dims,
+                in_channels=4096,
+                out_channels=256,
+                stride=1,
+                kernel_size=3,
+                norm_name="INSTANCE",
+                dropout=0,
+            ),
+            UnetOutBlock(
+                spatial_dims=spatial_dims,
+                in_channels=256,
+                out_channels=256,
+                dropout=0,
+            ),
+            Rearrange(
+                "b (c size_z size_y size_x) n_patch_z n_patch_y n_patch_x -> b c (n_patch_z size_z) (n_patch_y size_y) (n_patch_x size_x)",
+                n_patch_z=num_patches[0],
+                n_patch_y=num_patches[1],
+                n_patch_x=num_patches[2],
+                size_y=patch_size[1],
+                size_x=patch_size[2],
+                size_z=patch_size[0],
+            ),
+        )
+
+    def forward(self, img):
+        features = self.encoder(img)
+        features = self.decoder(features)
+        return features
+
+
+class IWMPredictor(torch.nn.Module):
```

Comment:
is this class used anywhere?

Reply:
yes, it's passed to the IWM class. I'll add configs to show how all the pieces come together